operator/builtin/input/file/reader.go

```diff
@@ -96,39 +97,73 @@ func (f *Reader) InitializeOffset(startAtBeginning bool) error {
 	return nil
 }
 
+type consumerFunc func(context.Context, []byte) error
+
 // ReadToEnd will read until the end of the file
 func (f *Reader) ReadToEnd(ctx context.Context) {
+	f.readFile(ctx, f.emit)
+}
+
+// ReadHeaders will read a files headers
+func (f *Reader) ReadHeaders(ctx context.Context) {
+	f.readFile(ctx, f.readHeaders)
+}
+
+func (f *Reader) readFile(ctx context.Context, consumer consumerFunc) {
 	if _, err := f.file.Seek(f.Offset, 0); err != nil {
 		f.Errorw("Failed to seek", zap.Error(err))
 		return
 	}
-
 	fr := NewFingerprintUpdatingReader(f.file, f.Offset, f.Fingerprint, f.fileInput.fingerprintSize)
 	scanner := NewPositionalScanner(fr, f.fileInput.MaxLogSize, f.Offset, f.fileInput.SplitFunc)
 
-	// Iterate over the tokenized file, emitting entries as we go
+	// Iterate over the tokenized file
 	for {
 		select {
 		case <-ctx.Done():
 			return
 		default:
 		}
-
 		ok := scanner.Scan()
 		if !ok {
 			if err := getScannerError(scanner); err != nil {
 				f.Errorw("Failed during scan", zap.Error(err))
 			}
 			break
 		}
-
-		if err := f.emit(ctx, scanner.Bytes()); err != nil {
-			f.Error("Failed to emit entry", zap.Error(err))
+		if err := consumer(ctx, scanner.Bytes()); err != nil {
+			// return if header parsing is done
+			if err == errEndOfHeaders {
+				// de-increment offset because scanner.Bytes()'s value was read but not used
+				f.Offset = f.Offset - 1
```

Comment:
I'm not sure we actually need to roll back the scanner's position. If we just return, then we skip updating the offset, which I think is sufficient. If I'm interpreting correctly, I'm pretty sure the scanner is not used again, and a new scanner is created the next time we try to read.

Reply:
Also, if we do need to rollback the scanner, I believe it would need to be by `len(scanner.Bytes()`

Reply:
Good point, I see that now. I tested before and after removing the de-increment and it works fine.